src/ripple/core/Pg.cpp

```diff
@@ -0,0 +1,1396 @@
+    {
+        std::stringstream ss;
+        ss << error_->first << ": " << error_->second;
+        return ss.str();
+    }
+    if (result_)
+        return "ok";
+
+    // Must be stopping.
+    return "stopping";
+}
+
+//-----------------------------------------------------------------------------
+
+/*
+ Connecting described in:
+ https://www.postgresql.org/docs/10/libpq-connect.html
+ */
+void
+Pg::connect()
+{
+    if (conn_)
+    {
+        // Nothing to do if we already have a good connection.
+        if (PQstatus(conn_.get()) == CONNECTION_OK)
+            return;
+        /* Try resetting connection. */
+        PQreset(conn_.get());
+    }
+    else  // Make new connection.
+    {
+        conn_.reset(PQconnectdbParams(
+            reinterpret_cast<char const* const*>(&config_.keywordsIdx[0]),
+            reinterpret_cast<char const* const*>(&config_.valuesIdx[0]),
+            0));
+        if (!conn_)
+            Throw<std::runtime_error>("No db connection struct");
+    }
+
+    /** Results from a synchronous connection attempt can only be either
+     * CONNECTION_OK or CONNECTION_BAD. */
+    if (PQstatus(conn_.get()) == CONNECTION_BAD)
+    {
+        std::stringstream ss;
+        ss << "DB connection status " << PQstatus(conn_.get()) << ": "
+           << PQerrorMessage(conn_.get());
+        Throw<std::runtime_error>(ss.str());
+    }
+
+    // Log server session console messages.
+    PQsetNoticeReceiver(
+        conn_.get(), noticeReceiver, const_cast<beast::Journal*>(&j_));
+}
+
+PgResult
+Pg::query(char const* command, std::size_t nParams, char const* const* values)
+{
+    // The result object must be freed using the libpq API PQclear() call.
+    pg_result_type ret{nullptr, [](PGresult* result) { PQclear(result); }};
+    // Connect then submit query.
+    while (true)
+    {
+        try
+        {
+            connect();
+            if (nParams)
+            {
+                // PQexecParams can process only a single command.
+                ret.reset(PQexecParams(
+                    conn_.get(),
+                    command,
+                    nParams,
+                    nullptr,
+                    values,
+                    nullptr,
+                    nullptr,
+                    0));
+            }
+            else
+            {
+                // PQexec can process multiple commands separated by
+                // semi-colons. Returns the response from the last
+                // command processed.
+                ret.reset(PQexec(conn_.get(), command));
+            }
+            if (!ret)
+                Throw<std::runtime_error>("no result structure returned");
+            break;
+        }
+        catch (std::exception const& e)
+        {
+            // Sever connection and retry until successful.
+            disconnect();
+            JLOG(j_.error()) << "database error, retrying: " << e.what();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
This essentially endless loop might cause the server to not be able to stop. Perhaps pass the PgPool::stop_ variable as reference to this function so that it can be checked?

Reply:
I pushed a commit that can be cherry-picked that fixes this: https://github.com/mtrippled/rippled/commit/be375ce611ca0290729187b8909c166bc00fc049